### DISCUSSION TITLE
Should there be one navigation - NavigationView?

### DISCUSSION THREAD

**Discussion by pomianowski** [lepoco/wpfui#329] [Ideas]
Hey guys,  
to make breaking changes as much as possible controlled I would like to ask if at this stage you think that there should be one navigation control - `NavigationView`.

Giving a bit of a wider perspective, when creating the current navigation, I did not follow WinUi but my own implementations created for the needs of other projects. While it made sense for a small scope of apps now I can see that there are too many use cases that I have not considered.

## Control

`NavigationView` would have a presenter - `Frame` or some other presenting control and a `Breadcrumb` included.
It should be a whole and not require additional controls. All this to keep everything as simple as possible.

For example:
```xaml
<ui:NavigationView
  PaneStyle="Compact"
  PaneDisplayMode="Left"
  TransitionType="FadeInWithSlide"
  IsBackEnabled="True"
  IsBackButtonVisible="True"
  IsBreadcrumbVisible="True"
  MenuItems="{Binding}"
  FooterMenuItems="{Binding}"
  Navigated="OnNavigated">
  <ui:NavigationView.Header>
    <!-- Alternate way to create a breadcrumb? -->
  </ui:NavigationView.Header>
  <ui:NavigationView.Breadcrumb>  <!-- Accept object or FrameworkElement ? -->
    <ui:Breadcrumb
      FontSize="12" />
  </ui:NavigationView.Breadcrumb>
  <ui:NavigationView.Content>
    <!--
      Custom content rendered above the content of the page,
      may be breadcrumb or whatever else you like
      maybe dialog or snackbar? -->
  </ui:NavigationView.Content>
</ui:NavigationView>
```

```c#
NavigationView.PaneDisplayMode = PaneDisplayMode.Top; // Top / Left
```
```c#
NavigationView.SelectedItem; // Instead of the currently used property - Current
```

Parameters named similarly to Win Ui `NavigationView` - `MenuItems` instead of `Items` and `FooterMenuItems` instead of `Footer`.

Instead of `NavigationStore` or `NavigationCompact` class, we could use `PaneStyle="Compact"` - Compact / Store / Fluent.

## Hamburger button
Hamburger button and Back button can be in title bar or the navigation itself. This is also a problem that should be solved. Should `TitleBar` implement a property like `Navigation="{Bind RootNav}"` or maybe a new `NavigationControls` control that could be placed in `TitleBar.Header`.

## Frame

And the issue of rendering the control itself. We can skip the whole `Frame` control related stuff and simply use `ContentPresenter` or `ScrollViewer`. This will reduce the total number of rendered elements but the effect will be the same. Anyway, the most common way to create page objects is to do it before entering them in `Frame`.

I would be grateful for your suggestions and comments.

**Comment by IvanDmitriev1** [comment]
What about #327 ?

**Comment by PierreLeGit** [comment]
Hello Lepo,

Current navigation is fine. However, the animation of the navigation banner suffers from slowing down when opening the pages, which unfortunately is not very pretty. I have already created issues that have remained unanswered. For me, this inconvenience should already be resolved. The breadcumd is a good idea but it must remain a separate control, whether we have the choice to use it or not. The Hamburger icon must be present if you decide to hide the current navigation banner. The back button in the title bar would be really nice!

**Reply by pomianowski** [reply]
> Hello Lepo,
> 
> Current navigation is fine. However, the animation of the navigation banner suffers from slowing down when opening the pages, which unfortunately is not very pretty. I have already created issues that have remained unanswered. For me, this inconvenience should already be resolved. The breadcumd is a good idea but it must remain a separate control, whether we have the choice to use it or not. The Hamburger icon must be present if you decide to hide the current navigation banner. The back button in the title bar would be really nice!

Solving the problem of navigation slowing is not trivial as everything has to happen in one thread. If you know a solution or have seen a implementation that provides decent performance when rendering a large number of controls, feel free to share.

**Reply by MajorXAML** [reply]
Well, I would suggest just using the navigation from ModernWpf. Of course by upgrading it to winui3.

**Reply by PierreLeGit** [reply]
> > Hello Lepo,
> > Current navigation is fine. However, the animation of the navigation banner suffers from slowing down when opening the pages, which unfortunately is not very pretty. I have already created issues that have remained unanswered. For me, this inconvenience should already be resolved. The breadcumd is a good idea but it must remain a separate control, whether we have the choice to use it or not. The Hamburger icon must be present if you decide to hide the current navigation banner. The back button in the title bar would be really nice!
> 
> Solving the problem of navigation slowing is not trivial as everything has to happen in one thread. If you know a solution or have seen a implementation that provides decent performance when rendering a large number of controls, feel free to share.

This wpf app showing perfect performances 
https://spiegelp.github.io/MaterialDesignExtensions/
I’ve tried to implement into wpfui but for me some knowledge or Time missing… so i’ll try It again. The disavantage is that navigation banner cannot be show in the Visual studio graphic editor because page are instanciate in array in code behind. However, the result is great with many kinds kind of content, especialy heavy.

**Reply by PierreLeGit** [reply]
Say me what do you think about that 😉

**Comment by PierreLeGit** [comment]
I've seen wpf project who use frame for content and the project contain just one page and the navigation baner is 100% fluid during page opening. This is maybe the right choise tu have reactive app.

**Comment by MajorXAML** [comment]
Of course NavigationView. 
By sticking to winui you can forget about writing documentation, because it's already there
